src/main/java/com/basho/riak/client/core/codec/TermToBinaryCodec.java

```diff
@@ -0,0 +1,294 @@
+        {
+            os.write_tuple_head(r.getCellsCount());
+            for (Cell c : r)
+            {
+                if (c == null)
+                {
+                    // NB: Null cells are represented as empty lists
+                    os.write_nil();
+                }
+                else
+                {
+                    os.write_any(c.getErlangObject());
+                }
+            }
+        }
+        os.write_nil();
+
+        return os;
+    }
+
+    private static QueryResult decodeTsResponse(byte[] response) throws OtpErlangDecodeException, InvalidTermToBinaryException
+    {
+        QueryResult result = null;
+
+        OtpInputStream is = new OtpInputStream(response);
+        final int msgArity = is.read_tuple_head();
+        // Response is:
+        // {'rpberrorresp', ErrMsg, ErrCode}
+        // {'tsgetresp', {ColNames, ColTypes, Rows}}
+        // {'tsqueryresp', {ColNames, ColTypes, Rows}}
+        final String respAtom = is.read_atom();
+        switch (respAtom)
+        {
+            case "rpberrorresp":
+                // TODO process error
+                assert (msgArity == 3);
+                break;
+            case "tsgetresp":
+            case "tsqueryresp":
+                assert (msgArity == 2);
+
+                final int dataArity = is.read_tuple_head();
+                assert (dataArity == 3);
+
+                final ArrayList<RiakTsPB.TsColumnDescription> columnDescriptions = parseColumnDescriptions(is);
+
+                final ArrayList<RiakTsPB.TsRow> rows = parseRows(is, columnDescriptions);
+
+                result = new QueryResult(columnDescriptions, rows);
+
+                break;
+            default:
+                // TODO GH-611 throw exception?
+        }
+
+        return result;
+    }
+
+    private static ArrayList<RiakTsPB.TsColumnDescription> parseColumnDescriptions(OtpInputStream is)
+            throws OtpErlangDecodeException
+    {
+        final int colNameCount = is.read_list_head();
+        final String[] columnNames = new String[colNameCount];
+        for (int colNameIdx = 0; colNameIdx < colNameCount; colNameIdx++)
+        {
+            final String colName = new String(is.read_binary(), StandardCharsets.UTF_8);
+            columnNames[colNameIdx] = colName;
+
+            final boolean isLastRow = colNameIdx + 1 == colNameCount;
+            if (isLastRow)
+            {
+                is.read_nil();
+            }
+        }
+
+
+        final int colTypeCount = is.read_list_head();
+        assert (colNameCount == colTypeCount);
+        final String[] columnTypes = new String[colTypeCount];
+
+        for (int colTypeIdx = 0; colTypeIdx < colTypeCount; colTypeIdx++)
+        {
+            final String colType = is.read_atom();
+            columnTypes[colTypeIdx] = colType;
+
+            final boolean isLastRow = colTypeIdx + 1 == colNameCount;
+            if (isLastRow)
+            {
+                is.read_nil();
```

Comment:
If we get an empty list then the nil isn't there, and it would throw an exception.